### ISSUE TITLE
Inactive and unmaintained clients

### ISSUE THREAD

**Issue by ivanhercaz** [tldr-pages/tldr#4044]
In Gitter arise an issue very interesting about the maintenance of the tldr-pages clients lists, "lists" rather than "list" because we have two lists, one in the README and another one in a wiki page. The user @principis address this issue asking about if someone should clean the clients lists, and after a conversation with @sbrl and me, we agree it would be interesting to address it.

The [original message](https://gitter.im/tldr-pages/tldr?at=5ebc6917496be6031dfe86b3) of @principis is the next:
> Should someone look into cleaning up the clients list? For example the TLDR Elixir Client is 6 years old, archived and still says "(binaries not yet available)". There are probably more clients which aren't being updated and/or following the client spec.

Recently I announce that I began to work in an Elixir client. The main reason I began this project is because the TLDR Elixir Client was inactive, archived and its development does not seem to continue (check [TLDR Elixir Client repository](https://github.com/edgurgel/tldr_elixir_client)).

There are probably another clients in this situation: inactives officially, like the TLDR Elixir Client, and apparently inactive or unmaintained. This second category seem a bit more subjective, but it probably needs a review.

Some questions to think about this issue may be:

- What should be the criteria to keep a client in the list?
- Should we accept any client in the list? Or just the ones which accomplish with client specification?
- Should we consider the time without activity in the repository? By development? By important issues without activity?

**Comment by ivanhercaz** [comment]
I am going to answer in a comment the above questions:

**What should be the criteria to keep a client in the list?**
I think this is the key of this issue. If we decide to clean the lists we should establish a criteria in which we would use to substantiate the decision. Now I have not a clear answer, thus I think this issue is useful.

**Should we accept any client in the list? Or just the ones which accomplish with client specification?**
Accomplish with client specification might be a good point to start. This even affect me, because ExTldr does not accomplish with the client specification and in this case it would not be in the client lists. But being honest, I think it is an objective criteria based mainly in the fact that if a client accomplish with the latest client specification or have backward compatibility, it should be in the lists.

**Should we consider the time without activity in the repository? By development? By important issues without activity?**
Sometimes the time without commits or releases is not a fact to drop something, because the reason may be that the client is already fine and it does not need any other change in its source code. In the other side, sometimes it may indicate lack of maintenance or inactivity of the project even if the repository is not archived. This is more obvious when the client repository has important issues without responses by the maintainers or possible solutions, or even pull requests stacked without being reviewed.

Of course, if a client repository is archived, I think it should not be in the lists. It does not mean only inactivity, but lack of continuation of the project too.

**Comment by zdroid** [comment]
This definitely should be implemented, but I'd keep a list of inactive/archived clients somewhere too.

**Comment by sbrl** [comment]
This is an excellent issue to raise.

I think that compliance with the client specification might be a good place to start. What about if a client hasn't made progress in conforming to the client specification in 6 months, then it's classed as old / unmaintained? We could have a second list of clients below the primary list that contains the old / unmaintained clients.

**Comment by ivanhercaz** [comment]
Totally agree with both of you, @zdroid and @sbrl. It would be interesting to keep a list of old/unmaintained/inactive/archived clients, although I will put a warning note in the top. This list could be only in a wiki page.

> What about if a client hasn't made progress in conforming to the client specification in 6 months, then it's classed as old / unmaintained?

That seems a really interesting option. If someone propose to add a client to this repository, but it does not accomplish with the client specification, but there are users working on the task, it should be enough.

Of course, this requieres for our part to review every X months if the clients still accomplish the client specification or if there have had progress with the client specification accomplishment.

**Comment by sbrl** [comment]
> Of course, this requieres for our part to review every X months if the clients still accomplish the client specification or if there have had progress with the client specification accomplishment.

True. But I guess any option here requires regular review.

**Comment by principis** [comment]
I think regular review shouldn't be required in the sense that, if a client supports the current specification it'll probably be fine for a while. On the other hand, every change to the client specification should be checked against the clients.

**Comment by MasterOdin** [comment]
Alternatively, could make it so that the list of clients is done with categories of specification supported, and put the onus on the client authors to make sure their client is in the right place on the list. For something like tldr, at a certain point, there's a good chance a client could be "done" for a given version that no further updates happening on it isn't a sign of "unsupported", just that there's nothing to be accomplished. Even an archived one is "unsupported", but if it is accomplishes everything in say spec 1.1, what does it matter?

A fun idea would be also that one could in-theory every time a client specification update happens, make a release on this repo, and have a github action workflow get triggered to run through the list and post an issue on each client's issue tracker that a new spec version has been released, and to update their client, and then to bump it in the list of supported clients.

**Comment by principis** [comment]
@MasterOdin The problem is, if no further updates are happening, it will mean that when a new client specification comes out (for example for the future page layout change), the client will be unusable.
I don't completely understand your category system. Because if 2.0 comes out, 1.1 clients probably just won't work anymore. 

> A fun idea would be also that one could in-theory every time a client specification update happens, make a release on this repo, and have a github action workflow get triggered to run through the list and post an issue on each client's issue tracker that a new spec version has been released, and to update their client, and then to bump it in the list of supported clients.

Something like that seems like a great idea. Maybe not with a release, because there currently are none and it may confuse people.

**Comment by principis** [comment]
Simple table with every client:

| Client        | Status      | Supports spec 1.2  |
| -------------- | ------------ | ------------------------- |
| [Alfred Workflow](https://github.com/cs1707/tldr-alfred) |  |    |
| [Albert Plugin](https://github.com/bergercookie/awesome-albert-plugins/tree/master/plugins/tldr_pages) |  Recent commits  |  N/A?  |
| [tldroid](https://github.com/hidroh/tldroid) |    |    |
| [tldr](https://github.com/raylee/tldr) |    |    |
| [tldr-bash-client](https://gitlab.com/pepa65/tldr-bash-client) |    |    |
| [C# client](https://github.com/principis/tldr-sharp) |  supported  |  yes  |
| [C client](https://github.com/tldr-pages/tldr-cpp-client) |    |    |
| [Chrome Extension](https://github.com/hill/tldr-chrome) |    |    |
| [Crystal client](https://github.com/porras/tlcr) |    |    |
| [Dart client](https://github.com/hterkelsen/tldr) |    |    |
| [Dash docset](https://github.com/Moddus/tldr-python-dash-docset) |    |    |
| [tldr-docker](https://github.com/nutellinoit/tldr-docker) |    |    |
| [ExTldr](https://github.com/ivanhercaz/extldr). |  developing  |  in progress  |
| [TLDR Elixir Client](https://github.com/edgurgel/tldr_elixir_client) |  Archived  |    |
| [Emacs client](https://github.com/kuanyui/tldr.el) |    |    |
| [github.com/pranavraja/tldr](https://github.com/pranavraja/tldr) |    |    |
| [4d63.com/tldr](https://4d63.com/tldr) |    |    |
| [github.com/elecprog/tldr](https://github.com/elecprog/tldr) |  supported  |  partially  |
| [github.com/isacikgoz/tldr](https://github.com/isacikgoz/tldr) |    |    |
| [tldr-man-page](https://github.com/freesuraj/TLDR) |    |    |
| [tldr-pages](https://github.com/mflint/ios-tldr-viewer) |    |    |
| [Haskell client](https://github.com/psibi/tldr-hs) |    |    |
| [Node.js client](https://github.com/tldr-pages/tldr-node-client) |    |    |
| [OCaml client](https://github.com/RosalesJ/tldr-ocaml) |    |    |
| [Perl5 client](https://github.com/shoichikaji/perl-tldr) |    |    |
| [PHP client](https://github.com/BrainMaestro/tldr-php) |    |    |
| [tldr-python-client](https://github.com/tldr-pages/tldr-python-client) |    |    |
| [tldr.py](https://github.com/lord63/tldr.py) |    |    |
| [R client](https://github.com/kirillseva/tldrrr) |    |    |
| [Ruby client](https://github.com/YellowApple/tldrb) |    |    |
| [Rust client](https://github.com/dbrgn/tealdeer) |    |    |
| [Vim Client](https://github.com/wlemuel/vim-tldr) |    |    |
| [Visual Studio Code extension](https://github.com/bmuskalla/vscode-tldr) |    |    |
| [tldr.jsx](https://github.com/ostera/tldr.jsx): http://tldr.ostera.io/ |    |    |
| [DistroWatch](https://distrowatch.com/dwres.php?resource=man-pages) |    |    |
| [tldr.ooops.me](https://tldr.ooops.me) |    |    |

I don't have time to test a lot of clients, but I filled in some I know.

**Comment by MasterOdin** [comment]
> @MasterOdin The problem is, if no further updates are happening, it will mean that when a new client specification comes out (for example for the future page layout change), the client will be unusable.

Maybe. For example, clients that support client spec 1.1 should still work just fine under 1.2. If the spec goes from 1.2 -> 2, then yeah, clients may break. However, I think for that, just wrap the wording on the page so that you've got "Clients that support Specification <version>" and then a line divide, and a note being like "Clients under here conform to an older version of the specification. As such, they may or may not work." and then continue on with "Clients that support Specification <version-1>", etc. Of course, this can get a bit wordy, so would probably move this into the wiki or something and not in the README?

This is more of just a way to make it less onus on the tldr organization to keep abreast of all clients, and put it on the client developers themselves.

**Comment by sbrl** [comment]
We could even fill in which client spec version clients support? But that might be too much work.

I really like the idea of automatically opening an issue when we make a release on this repo / update the client spec. That seems like a great idea!

**Comment by ldorigo** [comment]
Please, yes. As an end user I really don't care what my client is written in or to have 25 different options. I want an easy, functional way to install and access tldr pages.

**Comment by bl-ue** [comment]
I know no one probably is going to agree with this :), but I always have and do think that we shouldn't have a "Node.js" client or "Python" client but a "CLI" client, and a "Keypirinha" client, and a "web" client, etc.  Of course, it's much easier said than done, but having fewer clients makes it easier to make big changes such as #4120.

**Comment by sbrl** [comment]
@bl-ue that's an interesting idea, but even within the idea of a CLI client you'll have a job developing a universal client that works on all platforms: Windows, macOS, Linux, BSD, Solaris, Haiku, .... the list of OSes goes on an on. And with Go you also have to worry about CPU architectures. [This page](https://wiki.debian.org/SupportedArchitectures) is a favourite reference of mine for that. So I think there are definite advantages to having lots of different clients.

**Comment by bl-ue** [comment]
@sbrl

1. Does the tldr client need to be supported on any other platforms besides the major ones, which Go supports?  Most tools, including most tools documented in this tldr repo, don't support any other OSes.
2. If some people do use those operating systems, how likely is it that they would ever use tldr on them?
3. Why does the tldr client need to run on operating systems that many other tools don't?

>And with Go you also have to worry about CPU architectures.

That's a good point, but Go does support several of the architectures listed on that page you linked, see https://blog.golang.org/ports.

>So I think there are definite advantages to having lots of different clients.

I think so too, but not necessarily because of the OS/arch issue.  We need clients for iOS, Android, Web, Vim, Docker, Keypirinha, etc., and I don't really see any way to have tldr in those environments without multiple clients.

But the fact that we have so many CLI clients for such a simple documentation source and having to watch them seems to me impractical.

@ldorigo wrote:
>As an end user I really don't care what my client is written in or to have 25 different options.

which I agree with and have though the same thing since I realizd that there were multiple CLI clients.  The tldr-pages organization actually has 4 different CLI clients that all serve basically the same purpose!

@ldorigo was actually very accurate when he said 25 clients, because I looked and the readme and parsed out **24** different CLI clients...

**Comment by sbrl** [comment]
You raise some good points. We do have a reference client in Node.js over in the tldr-node-client repo. But writing a new client requires lots of effort and work - and also programmers with skills in Go for example if Go were to be picked.

In my opinion, I think it's far too late to change what's already happened.

Despite this, there is room for improvement. Some way of tracking compliance with the client spec would perhaps be useful, and keeping track of which clients support the latest version of the client spec. Notifying client authors about updates to the spec would be very useful too.

**Comment by MasterOdin** [comment]
There may be value though in at least having the major package managers of the big distributions / OS all use the same client though (which is ideally maintained by tldr?). A quick look at the ecosystem:

| OS        | tldr client |
|--------|----------------|
| Debian | https://github.com/psibi/tldr-hs#readme |
| macOS | https://github.com/tldr-pages/tldr-c-client |
| arch      | https://github.com/tldr-pages/tldr-python-client |
| fedora  | https://github.com/tldr-pages/tldr-python-client |
| Chocolatey |https://github.com/isacikgoz/tldr |

**Comment by mflint** [comment]
> Simple table with every client:
> Client 	Status 	Supports spec 1.2
> [tldr-pages](https://github.com/mflint/ios-tldr-viewer) 		

This is client for iOS.
Supported: yes
Supports spec 1.2: no (I really should do that)

**Comment by sbrl** [comment]
@mflint if you have any queries when you're working on implementing the new versions of the spec, do get in touch! We'd love to help. Quick link to the client spec changelog for convenience: <https://github.com/tldr-pages/tldr/blob/master/CLIENT-SPECIFICATION.md#changelog> (we're on v1.5 now, but the versions in between are relatively minor changes)

**Comment by bl-ue** [comment]
>There may be value though in at least having the major package managers of the big distributions / OS all use the same client though (which is ideally maintained by tldr?). A quick look at the ecosystem:
>
>| OS        | tldr client |
>|--------|----------------|
>| Debian | https://github.com/psibi/tldr-hs |
>| macOS | https://github.com/tldr-pages/tldr-c-client |
>| arch      | https://github.com/tldr-pages/tldr-python-client |
>| fedora  | https://github.com/tldr-pages/tldr-python-client |
>| Chocolatey |https://github.com/isacikgoz/tldr |

I completely agree with this.  In my opinion, an official client written in an interpreted language conflicts a lot with this goal, because in order to run the client Node.js, Python, etc. has to be installed

That's why I was advocating for Go (https://github.com/tldr-pages/tldr/issues/4044#issuecomment-753662999).  In my opinion, Go is not difficult to learn, being similar in many ways to common languages like JavaScript and Python.  In fact, of all of the languages that I've learned (which is several), Go has been the easiest for me to learn.  For most software developers learning Go should be a simple task.

The other benefit of Go is that it compiles to a single binaries for _AIX, **Android**, **Darwin (macOS)**, Dragonfly, Illumos, **iOS**, **Linux**, NetBSD, OpenBSD, Plan9, **Solaris**, **Windows**_ (and many different architectures for each), with no depencies or runtime necessary for running (and one unique thing about Go is that you can compile a binary for any platform from any platform, e.g. create a `.exe` for Windows on macOS 😉).  It's not a runtime and text source code bundled together into with a wrapper, like you might do with Python or Node.js.  It's actually compiled (albeit the binaries are a bit larger, e.g. a "Hello World" C program is 12 KB while a Go one is 1.3 MB.  The [`terraform`](https://github.com/hashicorp/terraform) binary is only 73 MB, smaller than most LLVM binaries.)

<details>
<summary>Supported Go target platforms</summary>

```console
$ go version
go version go1.16.2 darwin/amd64$ 
$ go tool dist list
aix/ppc64
android/386
android/amd64
android/arm
android/arm64
darwin/amd64
darwin/arm64
dragonfly/amd64
freebsd/386
freebsd/amd64
freebsd/arm
freebsd/arm64
illumos/amd64
ios/amd64
ios/arm64
js/wasm
linux/386
linux/amd64
linux/arm
linux/arm64
linux/mips
linux/mips64
linux/mips64le
linux/mipsle
linux/ppc64
linux/ppc64le
linux/riscv64
linux/s390x
netbsd/386
netbsd/amd64
netbsd/arm
netbsd/arm64
openbsd/386
openbsd/amd64
openbsd/arm
openbsd/arm64
openbsd/mips64
plan9/386
plan9/amd64
plan9/arm
solaris/amd64
windows/386
windows/amd64
windows/arm
```

</details>

Of course, you couldn't use a Go binary for a Web client (althought you can compile a Go program to JavaScript for WebAssembly), and you couldn't conveniently and conventionally write a Go iOS UI.  But Go has excellent community support for all kinds of CLI needs (e.g. the very popular CLI argument parsing library [`cobra`](https://github.com/spf13/cobra)), and of all of our CLI clients, the largest amount written in a single language is Go, as we have 4 Go clients.

These are just my thoughts.

**Comment by vladimyr** [comment]
> I completely agree with this. In my opinion, an official client written in an interpreted language conflicts a lot with this goal, because in order to run the client Node.js, Python, etc. has to be installed

Keep in mind that as long as clients are installed using respective package managers those runtimes might get installed automatically due to being listed as dependencies. Also in the world of Linux and macOS Python is already there.

> with no depencies or runtime necessary for running (and one unique thing about Go is that you can compile a binary for any platform from any platform, e.g. create a `.exe` for Windows on macOS ).

> It's actually compiled (albeit the binaries are a bit larger, e.g. a "Hello World" C program is 12 KB while a Go one is 1.3 MB.

That holds true from a regular user's perspective but isn't technically correct. Go binaries are rather large exactly because they bundle (statically link) that _missing_ runtime. FWIW cross-compilation is very easy with golang but most certainly isn't unique to it.

With that being said I fully agree that golang is a very practical choice in regards to getting things up and running quickly. :+1:

**Comment by sbrl** [comment]
Given the way the Debian folks package Go stuff, it might actually be a bad idea. From my (very limited) experience, they seem to package Go stuff like C binaries - i.e. dynamically linked to library files.

Despite this, I'd suggest that the client we pick for this @bl-ue is not the "one and only client", but one of our "reference clients", though even here that doesn't completely resolve the issue of unmaintained clients, because someone who knows Go would have to make a long-term commitment to keep it updated. I'm still absolutely against dropping the current community-maintained clients for all the same reasons I explained before.

Having said this, It does make sense that the single client is chosen to fulfil the role for every default package in package managers. Rust or Go or X language here would all be good for different reasons, but thinking about maintainability I would suggest that we pick a client that:

1. Is written in a language that lots of people know already, so that it's more statistically likely to be maintained
2. Is easy to package for all the target platforms.

Although my Python is rather rusty, I'd perhaps suggest that the Python client would be the best fit here, since some form of Python is installed by default on most systems.

Whichever client is picked, we'll want to review their compliance with the spec, and open issues / PRs to fix any issues found (might want to do this anyway with the existing reference clients).

**Comment by bl-ue** [comment]
Agreed—Python is good and simple **but** not installed by default on Windows :/

Rust I'd definitely say no to because it's not that intuitive to learn—Go on the other hand it quite easy to learn as it really borrows from Python/JavaScript a lot, and has "pointers" but not really C-style pointers, you can't use pointer arithmetic with them; they're mostly equivalent to references in C++, and I'd say passing a variable by reference vs. by value is pretty obvious (once explained) to any developer.

>I'm still absolutely against dropping the current community-maintained clients for all the same reasons I explained before.

After a little more time spent with the project I've come to this view as well and I longer advocate for the approach—say we do get a good language picked out and a nice official CLI client written...what about web, Android, iOS, etc.?

**Comment by bl-ue** [comment]
Back to the point—I'm compiling a detailed table of all of the clients listed on the readme and whether or not they're maintained and whether or note they support pages that use non-English punctuation (e.g. a Chinese colon at the end) for #5240.  I'll post it here when I'm done and then we can trim off the unmaintained ones.

**Comment by vladimyr** [comment]
@sbrl I follow your reasoning here but I would like to add that if there are multiple good candidates from a maintenance perspective, we should favor the one that is best from the user's perspective i.e. fastest. I'd argue that both Python and Node tick the former box but let's see how they compare in regular workflows using something like this for instance: https://github.com/dbrgn/tealdeer/issues/129#issuecomment-688378779

**Comment by vladimyr** [comment]
> Rust I'd definitely say no to because it's not that intuitive to learn—Go on the other hand it quite easy to learn...

I understand the sentiment and personally find golang easier to work with but let's be objective. This is pretty much a biased statement 🙃

**Comment by MasterOdin** [comment]
> Agreed—Python is good and simple but not installed by default on Windows :/

Python 3.5 is also not installed by default on macOS and maybe a number of LTS Ubuntu releases if we want to get pedantic. However, I disagree on that being a disqualifying criteria as installing the right version of python is extremely straight forward (same as node).

Additionally, using the benchmark above for it:

```
→ hyperfine 'tldr tar'
Benchmark #1: tldr tar
  Time (mean ± σ):      96.5 ms ±   5.9 ms    [User: 69.9 ms, System: 21.9 ms]
  Range (min … max):    86.4 ms … 113.6 ms    25 runs

→ hyperfine 'man tar'
Benchmark #1: man tar
  Time (mean ± σ):     103.7 ms ±   4.6 ms    [User: 100.9 ms, System: 19.1 ms]
  Range (min … max):    96.2 ms … 117.5 ms    24 runs
```

So 30ms behind tealdeer and faster than man, seems fine to me on speed. 


At the end of the day, I would argue the only real criteria of choice is purely on what do the most community maintainers prefer for a language. If only one person wants or knows rust and writes a client in rust, then god forbid they ever leave. Going from [JetBrains' Dev Ecosystem Survey](https://www.jetbrains.com/lp/devecosystem-2020/), JS and Python are good choices versus Go, but as @bl-ue indicates, Go isn't too bad on learning, but it's not clear to me that the majority of maintainers would work on it.

**Comment by vladimyr** [comment]
@MasterOdin Factual, nuanced, practical. 💯 👍

**Comment by sbrl** [comment]
@bl-ue see my comment [here](https://github.com/tldr-pages/tldr/pull/5240#issuecomment-812933519) regarding utf-8 support in clients.

**Comment by bl-ue** [comment]
Everyone, please see #6163.  It's my first big step towards pushing this issue and the problem regardling clients in general to a close.  I'd appreciate some input there.

I'm thinking that opening the issue(s) on a release is the best approach.  GH notifications are very light and ephemeral-ish, while issues are very clear and noticeable.  They also don't rely on the user watching our repository for releases.

We still need to decide on opening one issue _here_ or one issue on each client's repo.  For reasons I stated in #6163, I prefer the former.

**Comment by sbrl** [comment]
Oh, automatically opening an issue against client repos on releases with a summary of changes sounds like a great idea! Then client authors can close the issue when they've implemented the new features.

**Comment by xiongchiamiov** [comment]
Fyi, right now I think as a result of this discussion, the readme starts a chain of broken links.  The readme has

> There are also various other clients provided by the community, both for the command-line and for other platforms. For a comprehensive list of clients, head over to our [Wiki](https://github.com/tldr-pages/tldr/wiki/tldr-pages-clients).

which is a non-existent wiki page.  There _is_ https://github.com/tldr-pages/tldr/wiki/Minimum-specifications-for-tldr-pages-command-line-clients , but that says the long list of clients is in the readme and links back to the section that says they're in the wiki.

If the decision about how to store them is pending, then probably all these references to a list existing should probably be cleaned up so new folks are confused trying to find a list that doesn't exist.

**Comment by MasterOdin** [comment]
@xiongchiamiov This has been fixed. Not sure why @Metolinam did this?

```
commit 368c76b16f957d1dad5a0ccf3924b5cec5659aa8 (HEAD -> master, origin/master, origin/HEAD)
Author: Metolinam <103972608+Metolinam@users.noreply.github.com>
Date:   Thu May 26 20:44:29 2022 +0300

    Destroyed tldr pages clients (markdown)
```

**Comment by Managor** [comment]
Closing this as there doesn't seem to be any issues with the clients page. Please reopen if you disagree.

**Comment by kbdharun** [comment]
@Managor The conversation here isn't just about listing the clients on the client page alone it is about how we handle inactive and unmaintained clients that follow old client specifications (and aren't archived). IG we can continue the conversation in #10333 since some points are already addressed.
